pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,972 @@
+	// the one we keep when we resolve duplicates later.
+	logrus.Debugf("[%v] locating any DN-based Active Directory users", identifyAdUserOperation)
+	for _, user := range users.Items {
+		if !isAdUser(&user) {
+			logrus.Debugf("[%v] user '%v' has no AD principals, skipping", identifyAdUserOperation, user.Name)
+			continue
+		}
+		principalID := adPrincipalID(&user)
+		logrus.Debugf("[%v] processing AD User '%v' with principal ID: '%v'", identifyAdUserOperation, user.Name, principalID)
+		if isGUID(principalID) {
+			logrus.Debugf("[%v] '%v' does not appear to be a DN-based principal ID, taking no action", identifyAdUserOperation, principalID)
+			continue
+		}
+		dn, _, err := getExternalIDAndScope(principalID)
+		if err != nil {
+			logrus.Errorf("[%v] failed to extract DN from principal '%v', cannot process user: '%v'", identifyAdUserOperation, err, user.Name)
+			continue
+		}
+		if i, exists := knownDnWorkUnits[dn]; exists {
+			logrus.Debugf("[%v] user %v is DN-based (%v), and a duplicate of %v",
+				identifyAdUserOperation, user.Name, dn, usersToMigrate[i].originalUser.Name)
+			// Make sure the oldest duplicate user is selected as the original
+			if usersToMigrate[i].originalUser.CreationTimestamp.Time.After(user.CreationTimestamp.Time) {
+				usersToMigrate[i].duplicateUsers = append(usersToMigrate[i].duplicateUsers, usersToMigrate[i].originalUser)
+				usersToMigrate[i].originalUser = &user
+			} else {
+				usersToMigrate[i].duplicateUsers = append(usersToMigrate[i].duplicateUsers, &user)
+			}
+		}
+	}
+
+	return usersToMigrate, missingUsers, skippedUsers
+}
+
+func replaceGUIDPrincipalWithDn(user *v3.User, dn string) {
+	var principalIDs []string
+	for _, principalID := range user.PrincipalIDs {
+		if !strings.HasPrefix(principalID, activeDirectoryPrefix) {
+			principalIDs = append(principalIDs, principalID)
+		}
+	}
+	principalIDs = append(principalIDs, activeDirectoryPrefix+dn)
+	user.PrincipalIDs = principalIDs
+}
+
+func isAdUser(user *v3.User) bool {
+	for _, principalID := range user.PrincipalIDs {
+		if strings.HasPrefix(principalID, activeDirectoryPrefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func adPrincipalID(user *v3.User) string {
+	for _, principalID := range user.PrincipalIDs {
+		if strings.HasPrefix(principalID, activeDirectoryPrefix) {
+			return principalID
+		}
+	}
+	return ""
+}
+
+func localPrincipalID(user *v3.User) string {
+	for _, principalID := range user.PrincipalIDs {
+		if strings.HasPrefix(principalID, localPrefix) {
+			return principalID
+		}
+	}
+	return ""
+}
+
+func isDistinguishedName(principalID string) bool {
+	// Note: this is the logic the original migration script used: DNs have commas
+	// in them, and GUIDs do not. This seems... potentially fragile? Could a DN exist
+	// in the root of the tree (or perhaps be specified relative to a branch?) and thus
+	// be free of commas?
+	return strings.Contains(principalID, ",")
```

Comment:
I don't think you can really wind-up with a DN without a comma, but this case can be fabricated and would break things.  Nicholas thought about maybe doing the reverse logic (is this a GUID), which might be easier.

Reply:
The other approach is simply to check for an `=` which even a theoretical root level object (ie, a DN like `uid=test@example.com`) ought to still contain. I went with the flawed logic for parity with the original migration script, but I have serious doubts about shipping it upon reflection. We should use more robust logic to prevent sending the AD server a bunch of bad requests.